modules/flowable-rest/src/main/java/org/flowable/rest/service/api/runtime/task/TaskResource.java

```diff
@@ -199,39 +197,52 @@ public String getTaskForm(@ApiParam(name = "taskId") @PathVariable String taskId
     }
 
     protected void completeTask(Task task, TaskActionRequest actionRequest) {
-        Map<String, Object> variablesToSet = null;
-        Map<String, Object> transientVariablesToSet = null;
+        ScopedVariableContainerHelper scopedVariableContainerHelper = new ScopedVariableContainerHelper();
 
         if (actionRequest.getVariables() != null) {
-            variablesToSet = new HashMap<>();
             for (RestVariable var : actionRequest.getVariables()) {
                 if (var.getName() == null) {
                     throw new FlowableIllegalArgumentException("Variable name is required");
                 }
 
                 Object actualVariableValue = restResponseFactory.getVariableValue(var);
-                variablesToSet.put(var.getName(), actualVariableValue);
+                if (var.getVariableScope() != null) {
+                    if (RestVariable.RestVariableScope.LOCAL.equals(var.getVariableScope())) {
+                        scopedVariableContainerHelper.setVariableLocal(var.getName(), actualVariableValue);
+                    } else {
+                        scopedVariableContainerHelper.setVariable(var.getName(), actualVariableValue);
+                    }
+                } else {
+                    scopedVariableContainerHelper.setVariable(var.getName(), actualVariableValue);
+                }
             }
         }
 
         if (actionRequest.getTransientVariables() != null) {
-            transientVariablesToSet = new HashMap<>();
             for (RestVariable var : actionRequest.getTransientVariables()) {
                 if (var.getName() == null) {
                     throw new FlowableIllegalArgumentException("Transient variable name is required");
                 }
 
                 Object actualVariableValue = restResponseFactory.getVariableValue(var);
-                transientVariablesToSet.put(var.getName(), actualVariableValue);
+                if (var.getVariableScope() != null) {
```

Comment:
Check if this nested if can be simplified into a single if with a compound condition.

Reply:
Simplified with a compound condition